### PR TITLE
`get_datagrid()` for `{emmeans}` and `{marginaleffects}`

### LINKED ISSUE #921: `get_datagrid()` for `{emmeans}` and `{marginaleffects}`
Following https://github.com/easystats/bayestestR/issues/661, this could be useful for making the outputs more functional.

@strengejacke wdyt?

``` r
get_datagrid.emm_list <- get_datagrid.emmGrid <- function(x, ...) {
  suppressWarnings(s <- as.data.frame(x))
  data.frame(s)[,1:(which(colnames(s) == attr(s, "estName"))-1), drop = FALSE] 
}

myfit <- lm(mpg ~ factor(gear) + cyl, data = mtcars)

myemms <- emmeans::emmeans(myfit, pairwise ~ gear | cyl)
insight::get_datagrid(myemms)
#>      cyl gear      contrast
#> 1 6.1875    3             .
#> 2 6.1875    4             .
#> 3 6.1875    5             .
#> 4 6.1875    . gear3 - gear4
#> 5 6.1875    . gear3 - gear5
#> 6 6.1875    . gear4 - gear5
```

@vincentarelbundock is there perhaps a better way to extract this information?

``` r
get_datagrid.predictions <- get_datagrid.comparisons <- get_datagrid.slopes <- function(x, ...) {
  data.frame(x)[,1:(which(colnames(x) == "estimate")-1), drop = FALSE] 
}

myme <- marginaleffects::avg_slopes(myfit, variables = "cyl", by = "gear")
insight::get_datagrid(myme)
#>   term    contrast gear
#> 1  cyl mean(dY/dX)    3
#> 2  cyl mean(dY/dX)    4
#> 3  cyl mean(dY/dX)    5
```

``` r

myme <- marginaleffects::avg_predictions(myfit, variables = "cyl", by = "gear")
insight::get_datagrid(myme)
#>   gear
#> 1    4
#> 2    3
#> 3    5
```

``` r

myme <- marginaleffects::avg_comparisons(myfit, variables = "cyl", by = "gear")
insight::get_datagrid(myme)
#>   term contrast gear
#> 1  cyl mean(+1)    3
#> 2  cyl mean(+1)    4
#> 3  cyl mean(+1)    5
```

<sup>Created on 2024-09-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by mattansb**
Closes #921

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Type: Package
 Package: insight
 Title: Easy Access to Model Information for Various Model Objects
-Version: 0.20.4
+Version: 0.20.4.001
```
Comment (strengejacke) [inline DESCRIPTION:4]: Just use `.1` here.

File: tests/testthat/test-get_datagrid.R [hunk 2]
```diff
@@ -213,8 +213,92 @@ test_that("get_datagrid - models", {
...
+  mod2 <- lm(mpg ~ hp + factor(cyl), data = mtcars)
+
+  hp_vals <- c(50, 100)
+
+  em1 <- emmeans::emmeans(mod1, ~ cyl | hp, at = list(hp = hp_vals))
+  em2 <- emmeans::regrid(em1)
+  em3 <- emmeans::emmeans(mod2, ~ cyl | hp, at = list(hp = hp_vals))
+
+  res <- get_datagrid(em1)
+  expect_equal(res, get_datagrid(em2))
```
Comment (strengejacke) [inline tests/testthat/test-get_datagrid.R:232]: See lintrs. I suggest adding either (or both) `ignore_attr = TRUE` and optionally, `tolerance = ...`.

File: R/get_datagrid.R [hunk 3]
```diff
@@ -1,9 +1,13 @@
 #' Create a reference grid
 #'
 #' Create a reference matrix, useful for visualisation, with evenly spread and
-#' combined values. Usually used to make generate predictions using [get_predicted()].
-#' See this [vignette](https://easystats.github.io/modelbased/articles/visualisation_matrix.html)
+#' combined values. Usually used to make generate predictions using
+#' [get_predicted()]. See this
+#' [vignette](https://easystats.github.io/modelbased/articles/visualisation_matrix.html)
 #' for a tutorial on how to create a visualisation matrix using this function.
+#' \cr\cr
```
Comment (strengejacke) [inline R/get_datagrid.R:8]: No need for `\cr` here, markdown should detect empty lines.

File: R/get_datagrid.R [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
+#' Alternatively, these can also be used to extract the "grid" columns from
+#' objects generated by `emmeans` and `marginaleffects`.
```
Comment (strengejacke) [inline R/get_datagrid.R:10]: Standard package formatting (with `\pkg`) is usually bold, I think this is (mostly) consistent throughout the docs, so please use `*` or `**` for packages.

File: R/get_datagrid.R [hunk 5]
```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
 
 
 
+# Functions for emmeans/marginaleffects ---------------
 
-# helper -----------------
+#' @rdname get_datagrid
+#' @export
+get_datagrid.emmGrid <- function(x, ...) {
+  suppressWarnings(s <- as.data.frame(x))
```
Comment (strengejacke) [inline R/get_datagrid.R:665]: See lintr, surround expression inside `()` with `{}`.

File: R/get_datagrid.R [hunk 6]
```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
...
+get_datagrid.emm_list <- function(x, ...) {
+  k <- length(x)
+  res <- vector("list", length = k)
+  for (i in seq_len(k)) {
+    res[[i]] <- get_datagrid(x[[i]])
+  }
 
+  all_cols <- Reduce(lapply(res, colnames), f = union)
+  for (i in seq_len(k)) {
+    res[[i]][,setdiff(all_cols, colnames(res[[i]]))] <- NA
```
Comment (strengejacke) [inline R/get_datagrid.R:683]: See lintr, put space after comma.

File: R/get_datagrid.R [hunk 7]
```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
...
+  }
+  out <- do.call("rbind", res)
+
+  clear_cols <- colnames(out)[sapply(out, Negate(anyNA))] # these should be first
+  out[,c(clear_cols, setdiff(colnames(out), clear_cols)), drop = FALSE]
```
Comment (strengejacke) [inline R/get_datagrid.R:688]: same here

File: R/get_datagrid.R [hunk 8]
```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
...
+}
+
+#' @rdname get_datagrid
+#' @export
+get_datagrid.slopes <- function(x, ...) {
+  cols_newdata <- colnames(attr(x, "newdata"))
+  cols_contrast <- colnames(x)[grep("^contrast_?", colnames(x))]
```
Comment (strengejacke) [inline R/get_datagrid.R:695]: I think you can use `startsWith()` instead of grep here, which should be faster.

``` r
microbenchmark::microbenchmark(
  grepl("^Species?", colnames(iris)),
  startsWith(colnames(iris), "Species"),
  times = 1000
)
#> Unit: microseconds
#>                                   expr   min     lq      mean median     uq
#>     grepl("^Species?", colnames(iris)) 9.001 10.202 14.314188 14.001 16.701
#>  startsWith(colnames(iris), "Species") 1.900  2.201  3.807115  2.501  3.901
#>      max neval cld
#>   81.401  1000  a 
#>  957.100  1000   b
```

<sup>Created on 2024-09-03 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

File: R/get_datagrid.R [hunk 9]
```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
...
+  varname <- NA
+  original_target <- by
+
+  if (!is.null(by)) {
+    if (is.data.frame(x) && by %in% names(x)) {
+      return(data.frame(varname = by, expression = NA))
+    }
+
+    # If there is an equal sign
+    if (grepl("length.out =", by, fixed = TRUE)) {
```
Comment (strengejacke) [inline R/get_datagrid.R:723]: Not sure if the `by` string could also be `"length.out="`, without white space? If so, you may remove white space characters from `by` before proceeding? Maybe add a test for this edge case, too?

File: R/get_datagrid.R [hunk 10]
```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
...
+    } else if (grepl("=", by, fixed = TRUE)) {
+      parts <- trim_ws(unlist(strsplit(by, "=", fixed = TRUE), use.names = FALSE)) # Split and clean
+      varname <- parts[1] # left-hand part is probably the name of the variable
+      by <- parts[2] # right-hand part is the real target
+    }
+
+    if (is.na(by_expression) && is.data.frame(x)) {
+      if (is.na(varname)) {
+        format_error(
+          "Couldn't find which variable were selected in `by`. Check spelling and specification."
```
Comment (strengejacke) [inline R/get_datagrid.R:734]: plural s?

**Review by strengejacke — CHANGES_REQUESTED** [review]

File: tests/testthat/test-get_datagrid.R (hunk 2 above)
Comment (mattansb) [reply tests/testthat/test-get_datagrid.R:232]: Changed to `expect_identical()`

File: R/get_datagrid.R (hunk 3 above)
Comment (mattansb) [reply R/get_datagrid.R:8]: An empty lines bumps these to the details section.

File: R/get_datagrid.R (hunk 8 above)
Comment (mattansb) [reply R/get_datagrid.R:695]: But I need that optional `_` character :/

File: R/get_datagrid.R (hunk 9 above)
Comment (mattansb) [reply R/get_datagrid.R:723]: No idea what this is - I just organized that file by putting all utility functions at the end.

File: R/get_datagrid.R (hunk 3 above)
Comment (strengejacke) [reply R/get_datagrid.R:8]: Really? Strange... Anyway, then let's consider as resolved.

File: R/get_datagrid.R (hunk 8 above)
Comment (strengejacke) [reply R/get_datagrid.R:695]: Ok.

File: R/get_datagrid.R (hunk 9 above)
Comment (strengejacke) [reply R/get_datagrid.R:723]: Ah, that's not your code? Ok, I'll try and check later.

**Review by strengejacke — APPROVED** [review]

**Comment by mattansb** [comment]
Thanks @strengejacke !
